Service/src/main/java/org/gusdb/wdk/service/filter/RequestLoggingFilter.java

```diff
@@ -172,31 +210,79 @@ public void aroundWriteTo(WriterInterceptorContext context) throws IOException,
       LOG.error("Failure to write response body", e);
-      logRequestCompletion(httpStatus, "write failed", context::getProperty, context::removeProperty);
+      logRequestCompletion(httpStatus, "write failed", getAndRemoveProp);
       throw e;
     }
   }
 
-  private static void logRequestCompletion(int httpStatus, String bodyWriteStatus,
-      Function<String,Object> getter, Consumer<String> remover) {
-    // skip logging as requested
-    Boolean omitRequestLogging = (Boolean)getter.apply(OMIT_REQUEST_LOGGING_PROP_KEY);
-    remover.accept(OMIT_REQUEST_LOGGING_PROP_KEY);
+  private static Function<String,Object> toGetAndRemove(Function<String,Object> getter, Consumer<String> remover) {
+    return key -> {
+      Object obj = getter.apply(key);
+      remover.accept(key);
+      return obj;
+    };
+  }
+
+  private static void logRequestCompletion(int httpStatus, String bodyWriteStatus, Function<String,Object> getAndRemoveProp) {
+
+    // gather parameters for what action should be taken, then remove props from context object
+    Boolean omitRequestLogging = (Boolean)getAndRemoveProp.apply(OMIT_REQUEST_LOGGING_PROP_KEY);
+    ResponseLoggingAction action = (ResponseLoggingAction)getAndRemoveProp.apply(REQUEST_LOGGING_ACTION_PROP_KEY);
+    Path filePath = (Path)getAndRemoveProp.apply(REQUEST_BODY_FILE_PATH_PROP_KEY); // will be null if action is NONE
+    boolean isErrorStatus = httpStatus >= 500 && httpStatus < 600;
+
+    // if property not found, then this request went unmatched; can ignore.
+    if (omitRequestLogging == null) return;
+
+    if (action != ResponseLoggingAction.NONE) {
+
+      // decide whether and what to log
+      if (!omitRequestLogging && isLogEnabled()) {
+        String bodyText = !isErrorStatus ? "" : ", Request Body:\n" + readRequestBody(filePath);
+        LOG.log(LOG_LEVEL, "Request completed [" + httpStatus + "]" + bodyText);
+      }
 
-    // if property not found, then this request went unmatched; can ignore.  If present, omit as requested.
-    if (omitRequestLogging != null && !omitRequestLogging) {
-      LOG.log(LOG_LEVEL, "Request completed [" + httpStatus + "]; response body " + bodyWriteStatus);
+      // delete the temporary request body file
+      try {
+        Files.delete(filePath);
+      }
+      catch (IOException e) {
+        LOG.warn("Unable to delete temporary body file: " + filePath);
+      }
     }
   }
 
+  /**
+   * Reads the request body file but only up to the max number of logging bytes and returns as a String
+   *
+   * @param filePath path of temporary file
+   * @return possibly truncated content of request body
+   * @throws IOException 
+   */
+  private static String readRequestBody(Path filePath) {
+    try (Reader r = new BufferedReader(new FileReader(filePath.toFile()))) {
+      StringBuilder resultBuilder = new StringBuilder();
+      int count = 0;
+      int next;
+      while (((next = r.read()) != -1) && count < MAX_REQUEST_LOGGING_BYTES) {
```

Comment:
I think we can assume the file is written in UTF-8.  We can confirm with this:
```
java -XshowSettings 2>&1 | grep -i encoding
```

But yeah, it's tricky, because the file will be written in UTF-8, but if we're worried about memory, the internal JVM storage will be twice that (in most cases).  So is the setting the size in the file or what will occupy memory when we read it back in?  This is more an issue for the total request size than the logging truncation though- I think if we change it to "MAX_REQUEST_LOGGING_CHARS", it will be _mostly_ accurate given how we're measuring.